transposed/transposed.md

```diff
@@ -0,0 +1,265 @@
+
+---
+title: "Add transposed special cases for P2642 layouts"
+document: D????
+date: 2024/03/26
+audience: LEWG
+author:
+  - name: Mark Hoemmen
+    email: <mhoemmen@nvidia.com>
+toc: true
+---
+
+# Authors
+
+* Mark Hoemmen (mhoemmen@nvidia.com) (NVIDIA)
+
+# Revision history
+
+* Revision 0 to be submitted for the post-Tokyo mailing before 2024/04/16
+
+# Abstract
+
+We propose to change the C++ Working Paper
+so that `linalg::transposed` includes special cases
+for `layout_left_padded` and `layout_right_padded`.
+These are the two mdspan layouts proposed by P2642R6,
+which was voted into the C++ Working Draft at the Tokyo 2024 WG21 meeting.
+This change will make it easier for `linalg` implementations
+to optimize for these two layouts by dispatching
+to an existing optimized C or Fortran BLAS.
+Delaying this until after C++26 would be a breaking change.
+
+# Before and after comparison
+
+## Example
+
+The following example shows how this proposal
+changes the return type of `transposed`.
+
+```c++ 
+// optimized overload
+extern void some_algorithm(
+  mdspan<const float, dextents<size_t, 2>, layout_right_padded<dynamic_extent>> A_T,
+  mdspan<const float, dextents<size_t, 2>, layout_left_padded<dynamic_extent>> B,
+  mdspan<float, dextents<size_t, 2>, layout_left_padded<dynamic_extent>> C);
+
+template<class GenericFallBackLayout>
+void some_algorithm(
+  mdspan<const float, dextents<size_t, 2>, GenericFallBackLayout> A_T,
+  mdspan<const float, dextents<size_t, 2>, layout_left_padded<dynamic_extent>> B,
+  mdspan<float, dextents<size_t, 2>, layout_left_padded<dynamic_extent>> C)
+{
+  // ... slow generic code ...
+}
+
+void some_function(size_t N) {
+  vector<float> A_storage(4 * N * N);
+  vector<float> B_storage(4 * N * N);
+  vector<float> C_storage(4 * N * N);
+
+  // A, B, and C are each a 2N x 2N matrix.
+  auto mapping = layout_left::mapping{extents{2 * N, 2 * N}};
+  mdspan A{A_storage.data(), mapping};
+  mdspan B{B_storage.data(), mapping};
+  mdspan C{C_storage.data(), mapping};
+
+  // ... fill A and B with useful values ...
+
+  // A_00, B_00, and C_00 each view the upper left
+  // N x N submatrix of its "parent."
+  mdspan A_00 = submdspan(A, tuple{0, N}, tuple{0, N});
+  mdspan B_00 = submdspan(B, tuple{0, N}, tuple{0, N});
+  mdspan C_00 = submdspan(C, tuple{0, N}, tuple{0, N});
+
+  // Approval of P2642R6 added this to the C++ Working Draft.
+  static_assert(is_same_v<
+    decltype(A_00)::layout_type,
+    layout_left_padded<dynamic_extent>>);
+  static_assert(A_00.stride(0) == 1); // compile-time value
+
+  mdspan A_00_T = linalg::transposed(A_00);
+  some_algorithm(A_00_T, B_00, C_00);
+}
+```
```

Comment:
I think the example is not easily-digestible. I would leave out B and C as they are not related to the choice of taking the fallback. I think it might be compelling to show a second mdspan with the transposed extents and basically just a comment next to `some_algorithm` calls for both mdspans showing that the second calls the optimized version while the first one doesn't because it's a transposed layout

Reply:
@nmm0 Thanks for the feedback! : - )

It is a bit of a long example.  I mainly want to show how this proposal relates to P2642.

I could remove creation of separate `B_00` and `C_00` from `B` resp. `C` and just focus on `A_00`.  I would like to preserve the idea of calling an optimized algorithm, though.  What do you think?

Reply:
I've simplified the example a bit.  Please feel free to give more feedback if you like; thanks!